akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -141,7 +141,7 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with StageLoggi
         }
       case r: RstStreamFrame =>
         buffer.onRstStreamFrame(r)
-        multiplexer.cancelSubStream(r.streamId)
+        onReset(r.streamId)
```

Comment:
So, the problem is here that the side-effects are executed in the wrong order? I.e. we want to go to the `Closed` state but `multiplexer.cancelSubStream` accesses the state before it has changed.

Reply:
No, the problem was that this resetting logic was shared between the `Open` and `HalfClosedLocal` states. As far as I understand:

In `Open`, the old behavior was correct: `buffer.onRstStreamFrame(r)` made sure we closed the incoming request stream and `multiplexer.cancelSubStream(r.streamId)` made sure we closed the outgoing response.

However, in the scenario that we have already closed the outgoing response we are in `HalfClosedLocal` rather than `Open`. If we are in `HalfClosedLocal` and we receive a RST, we should still close the incoming request stream, but since the outgoing response stream was already closed we should not call `multiplexer.cancelSubStream(r.streamId)`.

Reply:
What I meant was, that the next line says, we should go to `Close` but that only applies after `cancelSubStream` has already run (and triggered the warning in `HalfClosedLocal`.

Reply:
I.e. `updateState` was called in a nested fashion which is basically not supported. The fix is fine, though :)

Reply:
Ah - that problem might still be there for the `Open` state, then? For a different PR though.

Reply:
It's not cleanly solved but what happens is that in the open state it is able to handle the `handleOutgoingEnded` without a warning (temporarily shifting the state to half closed).